pinot-spi/src/main/java/org/apache/pinot/spi/executor/DecoratorExecutorService.java

```diff
@@ -0,0 +1,163 @@
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+/**
+ * DecoratorExecutorService is an abstract class that provides a way to decorate an ExecutorService with additional
+ * functionality.
+ *
+ * TODO: Convert this class and its usages into an Executor instead of an ExecutorService
+ */
+public abstract class DecoratorExecutorService implements ExecutorService {
+  private final ExecutorService _executorService;
+
+  public DecoratorExecutorService(ExecutorService executorService) {
+    _executorService = executorService;
+  }
+
+  protected abstract <T> Callable<T> decorate(Callable<T> task);
+
+  protected abstract Runnable decorate(Runnable task);
+
+  protected <T> Collection<? extends Callable<T>> decorateTasks(Collection<? extends Callable<T>> tasks) {
+    return tasks.stream().map(this::decorate).collect(Collectors.toList());
+  }
+
+  @Override
+  public void shutdown() {
+    _executorService.shutdown();
+  }
+
+  @Override
+  public List<Runnable> shutdownNow() {
+    return _executorService.shutdownNow();
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return _executorService.isShutdown();
+  }
+
+  @Override
+  public boolean isTerminated() {
+    return _executorService.isTerminated();
+  }
+
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit)
+      throws InterruptedException {
+    return _executorService.awaitTermination(timeout, unit);
+  }
+
+  @Override
+  public <T> Future<T> submit(Callable<T> task) {
+    return _executorService.submit(decorate(task));
+  }
+
+  @Override
+  public <T> Future<T> submit(Runnable task, T result) {
+    return _executorService.submit(decorate(task), result);
+  }
+
+  @Override
+  public Future<?> submit(Runnable task) {
+    return _executorService.submit(decorate(task));
+  }
+
+  @Override
+  public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks)
+      throws InterruptedException {
+    return _executorService.invokeAll(decorateTasks(tasks));
+  }
+
+  @Override
+  public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks,
+      long timeout, TimeUnit unit)
+      throws InterruptedException {
+    return _executorService.invokeAll(decorateTasks(tasks), timeout, unit);
+  }
+
+  @Override
+  public <T> T invokeAny(Collection<? extends Callable<T>> tasks)
+      throws InterruptedException, ExecutionException {
+    return _executorService.invokeAny(decorateTasks(tasks));
+  }
+
+  @Override
+  public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout,
+      TimeUnit unit)
+      throws InterruptedException, ExecutionException, TimeoutException {
+    return _executorService.invokeAny(decorateTasks(tasks), timeout, unit);
+  }
+
+  @Override
+  public void execute(Runnable command) {
+    _executorService.execute(decorate(command));
+  }
+
+  public static abstract class BeforeAfter extends DecoratorExecutorService {
```

Comment:
This is unused? Is it just mean to be an example extensible implementation?

Reply:
Yes, it is unused, but it can help future implementations and is a good example for other custom implementations as well.